README.md

# Tradata

## Description

This project automates the collection of daily stock market data, including option chain, index prices, and FII DII data from multiple stock market websites thereby efficiently presenting it to the user in a clear and summarized format. By using this automation, the data collection process takes an average of **2 minutes and 6 seconds**, compared to **7 minutes and 35 seconds** it would take to gather the data manually,thereby resulting in a significant time saving of approximately **72.31%**.

## Issues Faced
* Adverstisements in several websites changed the **xpath** for several elements resulting in error, to tackle this I added *Adblock extension* in the selenium webdriver.

* Some websites detected selenium being used, and hence **denied** connection requestion,so I used *undetected chromedriver* to access all the websites.










Extension/CHANGELOG.txt
## Versions ##

UNRELEASED:

Version: 5.6.0, Date: 2023-05-01
- Continue to show the /update page to let users know about the AdBlock Premium
- Provide experimental Flattr API to some websites
- Expanded partner access to experimental allowlisting API

Version: 5.5.0, Date: 2023-04-14
- Show the /update page to let users know about the AdBlock Premium
- Added sending telemetry data to a new in-product messaging (IPM) server
- Added a new query string parameter to the un-install page which will provide information on whether the user has used the experimental allowlisting feature

Version: 5.4.2, Date: 2023-02-21
- Added a consent dialog when a user enables the AdBlock Premium Sync feature (only for Firefox users)

Version: 5.4.1, Date: 2023-02-12
- Fixed the on click event handler for the "Upgrade your AdBlock" CTA in the AdBlock popup menu
- Show the /update page to let users know about the AdBlock Premium

Version: 5.4.0, Date: 2023-02-02
- Modified or updated several JavaScript modules to work in a service worker environment in preparation for migration to manifest V3.
  The modified modules include:
  - survey & on page icon
  - telemetry
  - anonymous filter list data collection
  - ad blocking statistics ('Stats')
- We also updated the following in preparation for migration to manifest V3
  - the AdBlock Options page to use message passing
  - created a new common utilies module for use within a service worker
- Updated several dependencies used within AdBlock:
  - snippets to 0.6.0
  - adblockplusui to 3.16
  - web ext SDK to 0.7.2

Version: 5.3.3, Date: 2022-12-19
- Stop opening the /update page to let users know about the AdBlock Premium
- Added the Crotian locale to update an old Chrome web store Crotian listing

Version: 5.3.2, Date: 2022-11-18
- Show the /update page to let users know about the AdBlock Premium

Version: 5.3.1, Date: 2022-11-16
- Fixed an issue with the sending of telemetry
- Fixed an issue with the format of data being sent to the backup log server
- Stop showing the /update page to let users know about recent changes to AdBlock
- Updated the License information, git configuration, and README files

Version: 5.3.0, Date: 2022-10-21
- Fixed an issue with the Premium Sync feature
- Added a backup IP address telemetry process that will be used in case of a network issue
- Updated the eyeo web ext SDK to version 0.5.0
- In preparation for Manifest V3, the following changes where made:
  - converted modules that use the setTimeout & setInterval function to use the Alarm API
  - removed references to local storage, and instead use the extension storage API
  - removed any reference to jQuery in background scripts
  - converted AJAX requests to use the Fetch API
  - removed ability to subscribe to custom filter lists

Version: 5.2.0, Date: 2022-10-10
- Show the /update page to let users know about recent changes to AdBlock

Version: 5.1.3, Date: 2022-10-07
- Stop showing the /update page to let users know about the new AdBlock VPN

Version: 5.1.2, Date: 2022-09-06
- Show the /update page to let users know about the new AdBlock VPN

Version: 5.1.1, Date: 2022-08-29
- Replaced the abp-snippets git submodule with the @eyeo/snippets npm package and upgraded it to 0.5.2
- Cleaned up and fixed a couple of issues with the AdBlock package.json file

Version: 5.1.0, Date: 2022-08-22
- Removed the 'social media' links on the AdBlock Options page when the user is subscribed to the Anti Social Filter list
- Fixed an issue on the AdBlock Bug Report page with the 'steps' field not highlighted when left blank
- Fixed an issue with a missing error message when a user would edit their custom filters on the Customize tab of the AdBlock Options page
- Fixed an issue with anonymous filter list usage data incorrectly determining the filter list subscriptions for a filter list rule
- Updated the AdBlock On Page icon


Version: 5.0.5, Date: 2022-08-16
- Fixed an issue with telemetry possibly being sent prior to initialization being complete
- Updated the Snippets library to version 0.4.1

Version: 5.0.4, Date: 2022-07-15
- Fixed an issue with a missing entry in the locale messages file

Version: 5.0.3, Date: 2022-07-15
- Fixed a performance issue in the anonymous data collection, and image swap feature

Version: 5.0.2, Date: 2022-06-28
- Fixed an issue in the release build script
- Updated to the new eyeo ad blocking SDK.
- Updated translations

Version: 4.46.2, Date: 2022-05-15
- Continue to show /update page for those users that haven't seen the Spring 2022 update

Version: 4.46.1, Date: 2022-05-02
- Continue to show /update page for those users that haven't seen the Spring 2022 update

Version: 4.46.0, Date: 2022-04-18
- Added a AdBlock popup menu message about the AdBlock VPN
- Show the /update page to let users know what features have been updated in the last 6 months.
- Added checks to enhance content script messages security

Version: 4.44.0, Date: 2022-03-07
- Removed the logic to show the "New" badge text on extension update
- In preparation for manifest V3, and to help us better understand how our users use AdBlock, we've added
  - AdBlock Options Page Telemetry
  - filter list subscription information to the ping data

Version: 4.43.0, Date: 2022-01-25
- Fixed several issues on the AdBlock popup menu
- Removed the CryptoCurrency BitCoin Mining Protection filter list from the list of subscriptions added for new users
- Continue to show the message about Distraction Control on the AdBlock popup menu

Version: 4.42.0, Date: 2022-01-03
- Added a AdBlock popup menu message about Distraction Control
- Improved the Distraction Control help flow
- Made some minor improvements to the Distraction Control tab on the AdBlock options page
- Removed the functionality to open the /update page

Version: 4.41.0, Date: 2021-12-10
- Continue to show /update page for those users that haven't seen the Fall 2021 update
- Added a CTA to notify our users about the AdBlock VPN

script.py
import undetected_chromedriver as uc
from selenium import webdriver
from selenium.webdriver.chrome.options import Options
import time
import datetime
chrome_options = Options()
# chrome_options.headless = True
# chrome_options.add_argument("window-size=1536,864")
chrome_options.add_argument("--start-maximized")
chrome_options.add_argument('--load-extension=C:\\Users\\imdec\\Desktop\\YASH\\Projects\\Extension')
if __name__ == '__main__':
    # 1: Nifty Moneycontrol
    st1=0
    # 2: Nifty Investing
    st2=0
    # 3: Bank Nifty Moneycontrol
    st3=1
    # 4: Bank Nifty Investing
    st4=1
    # 5: FII DII
    st5=1
    # 6: Dow Jones
    st6=1
    # 7: Oilprice
    st7=1
    # 8: Put Call Ratio
    st8=1
    # 9: Option Chain
    st9=1
    driver = uc.Chrome(options=chrome_options)
    driver.set_page_load_timeout(15)
    driver.get('chrome://version/')
    time.sleep(0.2)
    try:
        driver.switch_to.window(driver.window_handles[1])
        driver.close()
    except:
        pass
    time.sleep(0.2)

    try:
        driver.switch_to.window(driver.window_handles[1])
        driver.close()
    except:
        pass
    driver.switch_to.window(driver.window_handles[0])

    while st1:
        try:
            driver.get('https://www.moneycontrol.com/indian-indices/nifty-50-9.html')
            time.sleep(10)
            try:
                driver.find_element_by_xpath("/html/body/div[15]/div[2]/div[3]/button[1]").click()
            except:
                pass
            driver.find_element_by_xpath("/html/body/div[9]/div/div[2]").screenshot("C:\\Users\\imdec\\Desktop\\YASH\\Projects\\Resources\\1.png")
            try:
                driver.find_element_by_xpath("/html/body/div[15]/div[2]/div[3]/button[1]").click()
            except:
                pass
            time.sleep(0.2)
            driver.find_element_by_xpath("/html/body/div[9]/div/div[5]/div[1]/div[1]/div[1]/div[2]/div").screenshot("C:\\Users\\imdec\\Desktop\\YASH\\Projects\\Resources\\2.png")
            try:
                driver.find_element_by_xpath("/html/body/div[15]/div[2]/div[3]/button[1]").click()
            except:
                pass
            time.sleep(2)
            driver.find_element_by_xpath("/html/body/div[9]/div/div[5]/div[1]/div[1]/div[1]/div[4]").screenshot("C:\\Users\\imdec\\Desktop\\YASH\\Projects\\Resources\\3.png")
            time.sleep(0.2)
            driver.find_element_by_xpath("/html/body/div[9]/div/div[5]/div[1]/div[1]/div[1]/div[9]").screenshot("C:\\Users\\imdec\\Desktop\\YASH\\Projects\\Resources\\4.png")
            driver.execute_script("window.scrollTo(0,20);")
            driver.find_element_by_xpath("/html/body/div[9]/div/div[5]/div[1]/div[3]/div[1]/div[1]").screenshot("C:\\Users\\imdec\\Desktop\\YASH\\Projects\\Resources\\5.png")
            driver.execute_script("window.scrollTo(0,20);")
            driver.find_element_by_xpath("/html/body/div[9]/div/div[5]/div[1]/div[3]/div[2]/div[1]").screenshot("C:\\Users\\imdec\\Desktop\\YASH\\Projects\\Resources\\6.png")
            break
        except:
            pass

    while st2:
        try:
            driver.get('https://in.investing.com/indices/s-p-cnx-nifty-components')
            time.sleep(2)
            driver.find_element_by_xpath("/html/body/div/div[4]/section[2]/div/section[2]/section[2]/div[1]/div/table/thead/tr/th[8]").click()
            driver.execute_script("window.scrollTo(0,610);")
            driver.find_element_by_xpath("/html/body/div/div[4]/section[2]/div/section[2]/section[2]/div[1]/div/table").screenshot("C:\\Users\\imdec\\Desktop\\YASH\\Projects\\Resources\\7.png")
            driver.find_element_by_xpath("/html/body/div/div[4]/section[2]/div/section[2]/section[2]/div[1]/div/table/thead/tr/th[8]").click()
            driver.find_element_by_xpath("/html/body/div/div[4]/section[2]/div/section[2]/section[2]/div[1]/div/table").screenshot("C:\\Users\\imdec\\Desktop\\YASH\\Projects\\Resources\\8.png")
            break
        except:
            pass

    while st3:
        try:
            driver.get('https://www.moneycontrol.com/indian-indices/bank-nifty-23.html')
            time.sleep(5)
            try:
                driver.find_element_by_xpath("/html/body/div[15]/div[2]/div[3]/button[1]").click()
            except:
                pass
            driver.find_element_by_xpath("/html/body/div[9]/div/div[2]").screenshot("C:\\Users\\imdec\\Desktop\\YASH\\Projects\\Resources\\9.png")
            time.sleep(0.2)
            driver.find_element_by_xpath("/html/body/div[9]/div/div[5]/div[1]/div[1]/div[1]/div[2]/div").screenshot("C:\\Users\\imdec\\Desktop\\YASH\\Projects\\Resources\\10.png")
            time.sleep(2)
            driver.find_element_by_xpath("/html/body/div[9]/div/div[5]/div[1]/div[1]/div[1]/div[4]").screenshot("C:\\Users\\imdec\\Desktop\\YASH\\Projects\\Resources\\11.png")
            time.sleep(0.2)
            driver.find_element_by_xpath("/html/body/div[9]/div/div[5]/div[1]/div[1]/div[1]/div[9]").screenshot("C:\\Users\\imdec\\Desktop\\YASH\\Projects\\Resources\\12.png")
            break
        except:
            pass

    while st4:
        try:
            driver.get('https://in.investing.com/indices/bank-nifty-components')
            time.sleep(2)
            driver.find_element_by_xpath("/html/body/div/div[4]/section[2]/div/section[2]/section/div[1]/div/table/thead/tr/th[8]").click()
            driver.execute_script("window.scrollTo(0,500);")
            driver.find_element_by_xpath("/html/body/div/div[4]/section[2]/div/section[2]/section/div[1]/div/table").screenshot("C:\\Users\\imdec\\Desktop\\YASH\\Projects\\Resources\\13.png")
            break
        except:
            pass

    while st5:
        try:
            driver.get('https://www.moneycontrol.com/stocks/marketstats/fii_dii_activity/index.php')
            time.sleep(2)
            try:
                driver.find_element_by_xpath("/html/body/div[15]/div[2]/div[3]/button[1]").click()
            except:
                pass
            driver.find_element_by_xpath("/html/body/section/div/div[4]/div[1]").screenshot("C:\\Users\\imdec\\Desktop\\YASH\\Projects\\Resources\\14.png")
            time.sleep(1)
            driver.find_element_by_xpath("/html/body/section/div/div[4]/div[1]/ul/li[2]/a").click()
            time.sleep(1)
            driver.find_element_by_xpath("/html/body/section/div/div[4]/div[1]").screenshot("C:\\Users\\imdec\\Desktop\\YASH\\Projects\\Resources\\15.png")
            time.sleep(1)
            driver.find_element_by_xpath("/html/body/section/div/div[4]/div[1]/div/div[2]/div[1]/div/ul/li[2]/a").click()
            time.sleep(1)
            driver.find_element_by_xpath("/html/body/section/div/div[4]/div[1]").screenshot("C:\\Users\\imdec\\Desktop\\YASH\\Projects\\Resources\\16.png")
            break
        except:
            pass
    while st6:
        try:
            driver.get('https://www.google.com/finance/quote/.DJI:INDEXDJX')
            time.sleep(1)
            driver.find_element_by_xpath("/html/body/c-wiz/div/div[4]/div/main").screenshot("C:\\Users\\imdec\\Desktop\\YASH\\Projects\\Resources\\17.png")
            break
        except:
            pass
    while st7:
        try:
            driver.get('https://oilprice.com/')
            time.sleep(2)
            driver.find_element_by_xpath("/html/body/div[2]/div[3]/div/div[1]").screenshot("C:\\Users\\imdec\\Desktop\\YASH\\Projects\\Resources\\18.png")
            break
        except:
            pass

    exp=3 #For thursday it is 3
    x=datetime.datetime.now()
    if x.weekday()<exp:
        x=x-datetime.timedelta(x.weekday())+datetime.timedelta(exp)
    else:
        x+=datetime.timedelta(7)
        x=x-datetime.timedelta(x.weekday())+datetime.timedelta(exp)
    hlpr=str(x.strftime("%d")+"-"+x.strftime("%m")+"-"+x.strftime("%Y"))
    print(hlpr)


    while st8:
        try:
            driver.get('https://www.niftytrader.in/nifty-put-call-ratio')
            time.sleep(3)
            driver.find_element_by_xpath("/html/body/div[1]/div/div/div[5]/div/div[1]/div[2]").screenshot("C:\\Users\\imdec\\Desktop\\YASH\\Projects\\Resources\\19.png")
            driver.find_element_by_xpath("/html/body/div[1]/div/div/div[5]/div/div[1]/div[2]/div[2]/div[2]/div/div/a[2]").click()
            time.sleep(4)
            driver.find_element_by_xpath("/html/body/div[1]/div/div/div[5]/div/div[1]/div[2]").screenshot("C:\\Users\\imdec\\Desktop\\YASH\\Projects\\Resources\\20.png")
            break
        except:
            pass

    time.sleep(1.5)

    exp=3 #For thursday it is 3
    x=datetime.datetime.now()
    if x.weekday()<exp:
        x=x-datetime.timedelta(x.weekday())+datetime.timedelta(exp)
    else:
        x+=datetime.timedelta(7)
        x=x-datetime.timedelta(x.weekday())+datetime.timedelta(exp)
    hlpr=str(x.strftime("%d")+"-"+x.strftime("%m")+"-"+x.strftime("%Y"))
    print(hlpr)
    while st9:
        try:
            driver.get('https://www.nseindia.com/option-chain')

            time.sleep(2)
            hlpr=str(x.strftime("%d")+"-"+x.strftime("%b")+"-"+x.strftime("%Y"))
            for i in range(1,100):
                print(i)
                strr='/html/body/div[10]/div[2]/section/div/div/div/div/div[1]/div[1]/div/div/div[2]/div[3]/div/select/option['+str(i)+']'
                print(driver.find_element_by_xpath(strr).get_attribute("value"),hlpr)
                if driver.find_element_by_xpath(strr).get_attribute("value")==hlpr:
                    driver.find_element_by_xpath(strr).click()
                    break
            time.sleep(1)

            curr = driver.find_element_by_xpath('/html/body/div[10]/div[2]/section/div/div/div/div/div[1]/div[1]/div/div/div[3]/div[1]/div[1]/span[2]').text
            curr=int(curr.replace(',','').replace('NIFTY ','')[0:-3])
            curr=curr-(curr%50)
            ans=curr+650
            for i in range(1000):
                iter = driver.find_element_by_xpath('/html/body/div[10]/div[2]/section/div/div/div/div/div[1]/div[1]/div/div/div[3]/div[2]/div/div/table[1]/tbody/tr['+str(i+1)+']/td[12]/a').text
                iter=int(iter.replace(',','')[0:-3])
                if (ans==iter):
                    driver.find_element_by_xpath('/html/body/div[10]/div[2]/section/div/div/div/div/div[1]/div[1]/div/div/div[3]/div[2]/div/div/table[1]/tbody/tr['+str(i+1)+']/td[12]/a').screenshot("C:\\Users\\imdec\\Desktop\\YASH\\Projects\\Resources\\21.png")
                    break
            time.sleep(0.5)
            driver.find_element_by_xpath("/html/body/div[10]/div[2]/section/div/div/div/div/div[1]/div[1]/div/div/div[3]/div[2]/div/div/table[1]").screenshot("C:\\Users\\imdec\\Desktop\\YASH\\Projects\\Resources\\21.png")
            driver.execute_script("window.scrollTo(0,0);")
            driver.find_element_by_xpath('/html/body/div[10]/div[2]/section/div/div/div/div/div[1]/div[1]/div/div/div[2]/div[1]/div/select/option[4]').click()
            time.sleep(0.3)
            for i in range(1,100):
                strr='/html/body/div[10]/div[2]/section/div/div/div/div/div[1]/div[1]/div/div/div[2]/div[3]/div/select/option['+str(i)+']'
                if driver.find_element_by_xpath(strr).get_attribute("value")==hlpr:
                    print(driver.find_element_by_xpath(strr).get_attribute("value"))
                    driver.find_element_by_xpath(strr).click()
                    break

            time.sleep(1)
            for i in range(1,100):
                strr='/html/body/div[10]/div[2]/section/div/div/div/div/div[1]/div[1]/div/div/div[2]/div[3]/div/select/option['+str(i)+']'
                if driver.find_element_by_xpath(strr).get_attribute("value")==hlpr:
                    print(driver.find_element_by_xpath(strr).get_attribute("value"))
                    driver.find_element_by_xpath(strr).click()
                    break

            time.sleep(1)

            curr = driver.find_element_by_xpath('/html/body/div[10]/div[2]/section/div/div/div/div/div[1]/div[1]/div/div/div[3]/div[1]/div[1]/span[2]').text
            curr=int(curr.replace(',','').replace('BANKNIFTY ','')[0:-3])
            curr=curr-(curr%100)
            ans=curr+1300
            for i in range(1000):
                iter = driver.find_element_by_xpath('/html/body/div[10]/div[2]/section/div/div/div/div/div[1]/div[1]/div/div/div[3]/div[2]/div/div/table[1]/tbody/tr['+str(i+1)+']/td[12]/a').text
                iter=int(iter.replace(',','')[0:-3])
                if (ans==iter):
                    driver.find_element_by_xpath('/html/body/div[10]/div[2]/section/div/div/div/div/div[1]/div[1]/div/div/div[3]/div[2]/div/div/table[1]/tbody/tr['+str(i+1)+']/td[12]/a').screenshot("C:\\Users\\imdec\\Desktop\\YASH\\Projects\\Resources\\22.png")
                    break

            time.sleep(0.5)
            driver.find_element_by_xpath("/html/body/div[10]/div[2]/section/div/div/div/div/div[1]/div[1]/div/div/div[3]/div[2]/div/div/table[1]").screenshot("C:\\Users\\imdec\\Desktop\\YASH\\Projects\\Resources\\22.png")
            break
        except:
            pass
    driver.quit()
    print("\n\n\n\n\n\n\n\n\n\nProgram Ended")